File: src/types/Notification.ts
export type NotificationCategory = 'Project' | 'Task' | 'Government' | 'AI' | 'System'

export interface NotificationLink {
  routeName: string
  params?: Record<string, string>
}

export interface AppNotification {
  id: string
  title: string
  message: string
  category: NotificationCategory
  date: string
  read: boolean
  link?: NotificationLink
}

export type NotificationGroupLabel = 'Today' | 'Yesterday' | 'Earlier'

export interface NotificationGroup {
  label: NotificationGroupLabel
  notifications: AppNotification[]
}

File: backend/app/schemas/notification.py
from datetime import datetime

from pydantic import BaseModel


class NotificationLinkOut(BaseModel):
    routeName: str
    params: dict[str, str] | None = None


class NotificationOut(BaseModel):
    id: str
    title: str
    message: str
    category: str
    date: datetime
    read: bool
    link: NotificationLinkOut | None = None

    @staticmethod
    def from_model(notification) -> "NotificationOut":
        link = (
            NotificationLinkOut(routeName=notification.link_route_name, params=notification.link_params)
            if notification.link_route_name
            else None
        )
        return NotificationOut(
            id=notification.notification_no,
            title=notification.title,
            message=notification.message,
            category=notification.category,
            date=notification.created_at,
            read=notification.read,
            link=link,
        )
